### DISCUSSION TITLE
Images broken v0.8.5 to v1.1.1

### DISCUSSION THREAD

**Discussion by Latro666** [dompdf/dompdf#2744] [Q&A]
Hi there

We are doing a provisional test of upgrading an older version 0.8.5 to 1.1.1 and it has had the result of images not loading in the PDF where they did before.

Images were used relative to the url generating the PDF e.g. /assests/5/image.jpg and in some cases the absolute server paths e.g. /var/home/sites/thissite/assests/5/image.jpg

Both via the use of the  img tag and via CSS background property, both seem to have stopped working via the upgrade?

I checked the release notes and can't find anything that noted as a breaking change between versions, especially regarding images?

example code:
https://jsfiddle.net/n4fyekh0/

**Comment by Latro666** [comment]
PHP side is pretty simple:
	
	$dompdf = new Dompdf();
	$dompdf->loadHtml($html);
	$dompdf->setPaper('A4', 'landscape');
	$dompdf->setBasePath($_SERVER['DOCUMENT_ROOT']);
	$dompdf->render();
	$dompdf->stream();


If we downgrade back to 0.8.5 (even including the newer package dependencies) it works again.

**Comment by bsweeney** [comment]
With newer versions of Dompdf images are restricted to the path(s) specified by the chroot option. You'll need to add the path for any external resources doing something like the following:

```
$dompdf = new Dompdf(["chroot" => ["/var/home/sites/thissite/assests"]]);
```

**Reply by Latro666** [reply]
Hi

Ah that explains it! poor release reading on my part for 0.8.6

_'     * Prevents dompdf from accessing system files or other files on the webserver.
     * All local files opened by dompdf must be in a subdirectory of this directory.
     * DO NOT set it to '/' since this could allow an attacker to use dompdf to
     * read any files on the server.  This should be an absolute path.'_

By absolute path is that from root or the public web folder e.g.
/var/home/sites/blah/public/myassets
or
/myassets

So this is the options array used on the constructor to dompdf. I'm assuming in the same clss you'd envoke the 'set_option($key, $value)' method e.g. $dompdf->set_option('chroot','/whatever'); if you wanted to amend this at any point?

**Reply by bsweeney** [reply]
It would be the full path on the filesystem, so " /var/home/sites/blah/public/myassets".

> So this is the options array used on the constructor to dompdf. I'm assuming in the same clss you'd envoke the 'set_option($key, $value)' method e.g. $dompdf->set_option('chroot','/whatever'); if you wanted to amend this at any point?

Correct. I set the chroot option in the constructor, but you can use the set_option method to change the value of chroot any time.

**Reply by Latro666** [reply]
Awesome, thanks for taking the time out of your day to help me Brian, have a great week.

**Reply by Latro666** [reply]
Ah spoke too soon. This isnt working for me still even with three test variants, does the directory need chmod permissions or something?

![Capture](https://user-images.githubusercontent.com/50453840/149963815-b6dd8f5b-8ea7-41a3-97d6-c0d9bdef3683.JPG)

	$dompdf = new Dompdf(["chroot" => [$_SERVER['DOCUMENT_ROOT'] . '/assets/clients/11']]);	
	$dompdf->setPaper('A4', $orientation);
	$dompdf->setBasePath($_SERVER['DOCUMENT_ROOT']);
	$dompdf->loadHtml($html);
	$dompdf->render();
	$dompdf->stream();
	
	HTML is
![Capture2](https://user-images.githubusercontent.com/50453840/149964141-70d84dbb-2e85-4eb3-8e12-79fecd452f26.JPG)

no joy

![Capture4](https://user-images.githubusercontent.com/50453840/149964154-2296220e-abd8-40c5-a681-29430d256de5.JPG)

**Reply by Latro666** [reply]
Image exists
![Capture5](https://user-images.githubusercontent.com/50453840/149964803-85520716-2b88-4753-b4b2-09662236dba4.JPG)

**Reply by Latro666** [reply]
Also tried
$dompdf = new Dompdf(["chroot" => [$_SERVER['DOCUMENT_ROOT'] . '/assets/clients/']]);	
$dompdf = new Dompdf(["chroot" => [$_SERVER['DOCUMENT_ROOT'] . '/assets/clients']]);	

As the method signature says sub directory so tried that above with html
src="/11/cert-image.jpg"
src="11/cert-image.jpg"

still nothing

**Reply by Latro666** [reply]
Even tried
$dompdf = new Dompdf(["chroot" => [$_SERVER['DOCUMENT_ROOT'] . '/assets/clients']]);

Html src as
src="/home/storm/sites/REDACTED/public/assets/clients/11/cert-image.jpg"

Trying to follow:
https://github.com/dompdf/dompdf/wiki/Usage#security-restrictions-for-local-files

**Reply by Latro666** [reply]
Ah it works if the chroot is **not passed** as an array but as a **string**. e.g.
$dompdf = new Dompdf(["chroot" => $_SERVER['DOCUMENT_ROOT']]);

Does this mean though that all file references in the HTML also have to be absolute? If so i have a lot of html to change as they were all saved as relative src attributes.

e.g.
![Capturex](https://user-images.githubusercontent.com/50453840/149969723-f378ec33-e9c8-4c0f-bc17-c5dd7365907e.JPG)

**Reply by bsweeney** [reply]
When you're using loadHtml Dompdf looks for the images relative to the executing file, not the document root. And paths with a leading slash are looked for relative to the root of the filesystem. So, yeah, not convenient if you need to share the code between web and Dompdf.

We have had requests to add a "document root" configuration so that paths with leading slashes start from the document root instead of the filesystem root. That is something we're looking into doing.

**Reply by Latro666** [reply]
Cool we'll have to stick to version 0.8.5 then if later versions effectively enforce absolute filesystem paths in HTML you pass in for images.

**Reply by bsweeney** [reply]
Bummer, some good updates since then. The 2.0 release will include updates to path handling and I'm thinking of including a new "document root" option which, when set, would be one of the allowed paths.

**Reply by Latro666** [reply]
Yea we run a certificate generator with a online and pdf preview. All the HTML works on both with versions using  0.8.5. I'd have to come up with some sort of run time checking and adjusting the HTML to add in the absolute file system path when in PDF mode now or change every single template with a shortcode system that replaces with either nothing or the path hmmm 

ofc 150,000 other jobs to do.

**Reply by Mellthas** [reply]
If the paths in HTML are all relative (without leading slash), then setting both the `chroot` option and the base path (via `$dompdf->setBasePath()`) to the document root should work in my experience.
